File: 287_find_the_duplicate_number/findDuplicate.go
package main

func findDuplicate(nums []int) int {
	n := len(nums)

	if n < 1 {
		return -1
	}

	slow, fast := nums[0], nums[nums[0]]

	for slow != fast {
		slow = nums[slow]
		fast = nums[nums[fast]]
	}

	fast = 0

	for slow != fast {
		slow = nums[slow]
		fast = nums[fast]
	}

	return slow
}

File: 287_find_the_duplicate_number/findDuplicate.ts
function findDuplicate(nums: number[]): number {
  let n = nums.length;

  if (n < 1) return -1;

  let slow = nums[0];
  let fast = nums[nums[0]];

  while (slow != fast) {
    slow = nums[slow];
    fast = nums[nums[fast]];
  }

  fast = 0;

  while (slow != fast) {
    slow = nums[slow];
    fast = nums[fast];
  }

  return slow;
}
